### DISCUSSION TITLE
Multiple Tenants in a Location?

### DISCUSSION THREAD

**Discussion by loregood** [netbox-community/netbox#10403] [Ideas]
We use Locations to register offices on a specific Site (building).

- Building 1 (Sites)
  - Office 1 (Locations)
  - Office 2
  - ..

We use a Tenant Group to register all employees, and would like to assign employees (Tenants) to specific offices (Locations).

- Employees (Tenant Groups)
  - Worker 1 (Tenant)
  - Worker 2
  - Worker 3
  - ..

We can assign "Worker 1" to the Location "Office 1" without any problems. **Works:**

- Worker 1 (`@Office 1`)

But if "Worker 2" and "Worker 3" is sharing "Office 2" we're out of luck. **Does not work:**

- Worker 2 (`@Office 2`)
- Worker 3 (`@Office 2`)

The issue is that you cannot assign multiple Tenants to a Location ("There can be only one.."). You could create a Tenant Group (like "Office 2"), but we can't have "Worker 2" and "Worker 3" assigned to both the Employees Tenant Group and the "Office 2" Tenant Group simultaneously. And you can't assign only a Tenant Group to an endpoint.

I haven't found a good way to address this issue in v3.3.4. We currently have "Desk 1" and "Desk 2" assigned as Locations in "Office 2" and then assign "Worker 2" to Location "Desk 1" and so on..

- Building 1 (Sites)
  - Office 2 (Locations)
    - Desk 1 (Location), Tenant = `Worker 2`
    - Desk 2 (Location), Tenant = `Worker 3`

But it kinda feels (and looks) like a hack. I think it would be nicer if "Office 2" could contain a list of Tenants in the office.

- Building 1 (Sites)
  - Office 2 (Location), (Tenant = `Worker 2`, Tenant = `Worker 3`)

The above is similar to "Possible to assign multiple tenants to Site? [#10159](https://github.com/netbox-community/netbox/discussions/10159)", but for Locations instead of a Site.

I'm not sure if "a list" of Tenants assigned to a Locations is the best solution? It would/should then also be possible in all the other places where you can assign Tenant(s) for consistency. It might also introduce UI challenges to multiple standard views, like (Sites, Locations..) where Tenant currently is only a single item (not a list of items).

Another solution could be to allow a (specific) Tenant to be assigned to multiple Tenant Groups simultaneously -- AND allow a Tenant Group to be assigned to a specific endpoint (like Locations). It might be less work to implement and maybe more useful generally? But more work to configure in NetBox and harder to explain to newcomers in our specific example above (I think).

I would like to hear others thoughts/suggestions before creating an Issue (or not).

**Comment by candlerb** [comment]
I think you are using 'Tenant' in a way other than it was intended.  'Tenant' is intended to mean "customer" - someone who rents space/buys service and pays the bill - and is usually a company, not an individual.

The ability to assign a Tenant to a Site or Location is for cases where that Location is dedicated to a single customer: e.g. a Location could be a cage or room containing multiple racks, and the entire cage is purchased by one customer.  That's effectively what you've done with your "Desk" Location proposal.

Assigning the Location to multiple customers isn't supported, because you can assign the Racks and/or Devices within that Location to specific customers instead.

For what you are trying to do, "Contacts" is closer (Netbox v3.1+). You'll have the same problem, in that one Contact can't be a member of multiple Contact Groups.  However, you *can* assign multiple contacts one by one to a Location, and you *can* assign the same Contact to multiple Locations. You could have a Contact Group for "Employees", and you can also nest Contact Groups, which might be useful for departments or geographical areas.

Having said all that, Netbox isn't really intended to be an employee management system :-)

**Reply by loregood** [reply]
Thank you for thoughtful reply.

I agree that creating employees as Tenants might be unusual, but I do think that, as NetBox gets adopted by "traditional" companies IT departments (not Internet and other Service Providers) the "customer" might in some cases be employees.

We assign each employee their own VLAN and separate IPv4/IPv6 prefixes to isolate Ethernet and containing malware from spreading to other employees (no traffic is allowed inbound on all employees VLANs). We use NetBox to assign VLANs and Prefixes (and a lot more) - and Ansible to configure all our switches and OpenBSD firewalls (hundreds of interfaces and pf-rules) with NetBox as a dynamic inventory (using only information from NetBox). Ensuring our documentation is always correct.

Since all cables, rack patch panels and office patch panels are already configured (and connected) in NetBox, we know which ports to configure when an employee moves to "Office 2" and connects to "Office Patch Panel 42". Our IT staff can see everything assigned to an employee from VLANs, Prefixes, Devices, Cables, physical Location and yes even Contact information from a single view in NetBox (Tenant -> Employee).

It might be unusual, but I do think it's within the spirit of what NetBox is designed to handle.

Let's say you work at a company that does coworking spaces for a living and you want to use NetBox to document your infrastructure. You have multiple offices (Locations), some small assigned only to a specific customer and some large shared by multiple customers. You now have the same problem as we have, but with customers instead of employees.. :-)

**Comment by joaolucasmacedo** [comment]
You can also create a custom field for contacts pointing a location object, so you can associate a specific location to each contact.

**Reply by candlerb** [reply]
You can, although I'm not sure why you'd want to do that when you can directly associate a Location with one or more Contacts already.

**Reply by joaolucasmacedo** [reply]
I see. I meant custom fields, by the way.
I think it is just a way to tie thing up, as he wants to give Netbox a people management looking ;-)

**Reply by loregood** [reply]
Thank you for the suggestion. We already use Contact information for each Tenant (employee).

I was not aware that you could associate a Location with multiple Contacts already. I will look into that, if it's supported on our production NetBox (v3.2.9).

**Reply by loregood** [reply]
I have now assigned multiple Contacts to a Location. It works!  But it's a worse solution than assigning a Desk Location for our use case. You can't see who is occupying an office from the Location view, and you can't see an employees Location from the Tenant view.

**Reply by candlerb** [reply]
> You can't see who is occupying an office from the Location view

I don't understand. When I browse to a Location, e.g. `/dcim/locations/9/`, then I see a list of contacts in the right-hand pane:

![image](https://user-images.githubusercontent.com/44789/191260083-b1d2bf6f-7769-4bb9-a5be-4919a78f7c15.png)

Tested with Netbox v3.2.9.

Maybe the issue is that you have *nested* locations, and contacts set on the child are not seen on the parent, or vice versa?

Or perhaps it's a permissions problem?

**Reply by loregood** [reply]
I meant from the `/dcim/locations` view (sorry, if I was unclear). There we can see all our offices on a specific floor (we use "floor #".. as a Parent Location on a Site) and employee initials for each Office (because the Tenant = employee). 

Simpel example below.
<img width="1107" alt="Screen Shot 2022-09-20 at 14 59 00" src="https://user-images.githubusercontent.com/2863424/191264365-6c55d97f-4788-4263-b942-1b728004a2d0.png">

**Reply by loregood** [reply]
The other "view" that is really nice for IT staff is this one. Here they have all the information they need about a specific employee, and if they have to help them "in person" - they can just click the Location and know where to go.

<img width="1125" alt="Screen Shot 2022-09-20 at 15 04 36" src="https://user-images.githubusercontent.com/2863424/191265461-d7472e68-40e6-4b7b-9286-d92502cff69f.png">

**Reply by candlerb** [reply]
There is a many-to-many relationship between contacts and locations. Hence if you view a *specific* location, you'll see all the contacts; and if you view a specific contact, you'll see all the locations (and other objects associated to that contact).

If you're asking for an extra column in the Locations listing view, which shows *all* the contacts for that location, then that's doable.  (It would be a feature request).

However, I think you're asking for many-to-many between tenants and location, and IMO this is the wrong thing to do - not that my opinion carries any weight, I'm just another user. Tenants are primarily related to assets (e.g. devices, virtual machines, IP addresses etc) which have a single tenancy.  If a location has multiple tenants, it's because it contains multiple assets which belong to various tenants - and that's already modelled.

As you already suggested, you can have PC 1 + phone 1 + printer 1 and/or location Desk 1 all with tenant "Worker 1".  But I still think that's not really in the spirit of Netbox "tenants".  As I said before, a Tenant is typically a company who is your customer.  A human being is a Contact.

Perhaps what you really want is an Employee object, in which case you can model that yourself using a Plugin.

**Reply by loregood** [reply]
Thank you for taking so much time to reply to my comments. You have convinced me, that "Multiple Tenants in a Location"  might be a bad idea - at least for the majority of NetBox users (currently). I still think that as NetBox's gets adopted by IT departments in "traditional" companies, the Tenancy attributes will be used to represent internal "ownership" like employees and departments - and Contacts will be used to "get in touch" (not specifically to represent a human). We currently also use Contacts to represent various Ticket queues. So if something is wrong, or you need help with a specific Device, the Ticket queue (for that service) is the primary contact.

And yes, it was a many-to-many relationship between Contacts and Locations I was asking about.

**Reply by candlerb** [reply]
> I still think that as NetBox's gets adopted by IT departments in "traditional" companies, the Tenancy attributes will be used to represent internal "ownership" like employees and departments

Certainly. Here, in a company with multiple core services run by different teams, I use "Tenant" to associate a device or a VM with the responsible team.  (Then I get Prometheus labels with the tenancy, and use it to do alert routing - which is quite cool.  I don't currently use Contacts on the Tenant for this, although it would be possible).

However, it still remains the case that a single Device, or VM, or IP address, or Location can only be associated with a single Tenant.  That works OK for me, and if I needed to do anything cleverer, I'd use Tags.

> And yes, it was a many-to-many relationship between Contacts and Locations I was asking about.

That already exists. Is there something specific about this that you'd like to see changed?

P.S. I *can* see a use case for allowing one Contact to be a member of multiple Contact Groups, and also to be able to attach a Contact Group directly to an object.  I believe that's been requested already.  In that case, a "Contact Group" would model a "Team" quite nicely, as well as any other arbitrary grouping of people, e.g. for alert routing or whatever.

**Reply by loregood** [reply]
> And yes, it was a many-to-many relationship between Contacts and Locations I was asking about.

Argh, I meant a many-to-many relationship between Tenants and Location..

**Comment by loregood** [comment]
Since it's already possible to assign multiple Contacts to a Location, I would suggest to implement that for Tenants also.. :-D
